carma_cooperative_perception/src/msg_conversion.cpp

```diff
@@ -72,6 +136,59 @@ auto heading_to_enu_yaw(const units::angle::degree_t & heading) noexcept -> unit
   return units::angle::degree_t{std::fmod(-(remove_units(heading) - 90.0) + 360.0, 360.0)};
 }
 
+auto enu_orientation_to_true_heading(double yaw, const lanelet::BasicPoint3d& obj_pose,
+const std::shared_ptr<lanelet::projection::LocalFrameProjector>& map_projection) noexcept -> units::angle::degree_t
+{
+  // Get object geodetic position
+  lanelet::GPSPoint wgs_obj_pose = map_projection->reverse(obj_pose);
+
+  // Get WGS84 Heading
+  gsl::owner<PJ_CONTEXT *> context = proj_context_create();
+  gsl::owner<PJ *> transform = proj_create(context, map_projection->ECEF_PROJ_STR);
+  units::angle::degree_t grid_heading{std::fmod(90 - yaw + 360, 360)};
+
+  const auto factors = proj_factors(transform, proj_coord( proj_torad(wgs_obj_pose.lon), proj_torad(wgs_obj_pose.lat), 0, 0));
+  units::angle::degree_t grid_convergence{proj_todeg(factors.meridian_convergence)};
+
+  auto wgs_heading = grid_convergence + grid_heading;
+  
+  proj_destroy(transform);
+  proj_context_destroy(context);
+  
+  return wgs_heading;
+}
+
+// determine the object position offset in m from the current reference pose in map frame and external object pose
+auto calc_relative_position(const geometry_msgs::msg::PoseStamped & source_pose,
+  const carma_v2x_msgs::msg::PositionOffsetXYZ & position_offset) noexcept -> carma_v2x_msgs::msg::PositionOffsetXYZ
+{
+  carma_v2x_msgs::msg::PositionOffsetXYZ adjusted_offset;
+
+  adjusted_offset.offset_x.object_distance = position_offset.offset_x.object_distance - source_pose.pose.position.x;
+  adjusted_offset.offset_y.object_distance = position_offset.offset_y.object_distance - source_pose.pose.position.y;
+  adjusted_offset.offset_z.object_distance = position_offset.offset_z.object_distance - source_pose.pose.position.z;
+  adjusted_offset.presence_vector |= carma_v2x_msgs::msg::PositionOffsetXYZ::HAS_OFFSET_Z;
```

Comment:
The `presence_vector` should be zeroed out initially to avoid undefined initial value issues.